### PR TITLE
Feature/memory cache

### PR DESCRIPTION
**Pull request by tofixx**
keep copy of cached files in memory and cleanup after expiration

### CODE REVIEW

File: middleware/fileCache.js [hunk 1]
```diff
@@ -1,18 +1,55 @@
...
+    let cache = new Set();
+    fs.readdirSync(this.filePath).forEach((file) => {
+      cache.add(file);
+    });
+    console.log('cache contains currently', cache.size, 'elements');
+    return cache;
   }
 
   add(filename, format, data) {
+    this.cache.add(filename);
```
Comment (franticBIT) [inline middleware/fileCache.js]: Should only be added, if file was successfully written.

File: middleware/fileCache.js [hunk 2]
```diff
@@ -1,18 +1,55 @@
...
 class FileCache {
   constructor() {
     this.filePath = config.get('Caching.Imagepath');
     if (!fs.existsSync(this.filePath)) {
       fs.mkdir(this.filePath);
     }
+    this.cache = this.loadCache();
+    let cleanupCronInterval = config.get('Caching.CleanupCronInterval');
+    let expirationTimeSeconds = config.get('Caching.Expires');
+    let fileCache = this;
```
Comment (franticBIT) [inline middleware/fileCache.js]: Variable seems superfluous.

File: middleware/fileCache.js [hunk 3]
```diff
@@ -22,19 +59,35 @@ class FileCache {
...
 
   exists(hash) {
-    return fs.existsSync(this.filePath + hash);
+    return this.cache.has(hash);
   }
 
   remove(hash) {
     if (fs.existsSync(this.filePath + hash)) {
-      return fs.unlinkSync(this.filePath + hash);
+      this.cache.delete(hash);
```
Comment (franticBIT) [inline middleware/fileCache.js]: Should also only be done, if delete was successful.

File: middleware/fileCache.js [hunk 4]
```diff
@@ -22,19 +59,35 @@ class FileCache {
...
+      fs.unlinkSync(this.filePath + hash);
```
Comment (franticBIT) [inline middleware/fileCache.js]: Why not async?

File: package.json [hunk 5]
```diff
@@ -22,6 +22,7 @@
     "object-hash": "1.1.8",
     "pug": "~2.0.0-rc.1",
     "sharp": "~0.17.3",
+    "cron": "~1.2.1",
```
Comment (franticBIT) [inline package.json]: Please order alphabetically. 😀

File: middleware/fileCache.js [hunk 6]
```diff
@@ -1,18 +1,55 @@
...
+    this.cache = this.loadCache();
+    let cleanupCronInterval = config.get('Caching.CleanupCronInterval');
+    let expirationTimeSeconds = config.get('Caching.Expires');
+    let fileCache = this;
+    if (cleanupCronInterval && expirationTimeSeconds) {
+      try {
+        this.cleanupJob = new cron.CronJob(
+          cleanupCronInterval,
+          function () {
+            console.log('cache cleanup starts... with current cache size', fileCache.cache.size);
```
Comment (franticBIT) [inline middleware/fileCache.js]: Do we have a proper logging tool?

File: middleware/fileCache.js (hunk 2 above)
Comment (tofixx) [inline middleware/fileCache.js]: which one? cleanupCronInterval defines when the cleanup is scheduled while expirationTimeSeconds defines which data is outdated.

File: middleware/fileCache.js (hunk 6 above)
Comment (tofixx) [inline middleware/fileCache.js]: #56

File: middleware/fileCache.js (hunk 2 above)
Comment (franticBIT) [inline middleware/fileCache.js]: `fileCache`

Comment (tofixx) [inline middleware/fileCache.js]: i think the this reference is required later and no more available inside of cron.CronJob

**Comment by tofixx** [comment]
@franticBIT changed the add and removal to run asyncronously

File: middleware/fileCache.js (hunk 2 above)
Comment (franticBIT) [inline middleware/fileCache.js]: Ah... you are right, since `function` opens a new closure. Might work with arrow syntax. 😬

**Pull request merged**